### PR TITLE
Add EitherCart and rename Yoke ZCF constructors

### PR DESCRIPTION
**Pull request by sffc**
I put two changes in this PR.  Let me know if you want me to separate them.

1. New type `EitherCart` that allows for ergonomic polymorphic carts
2. Renamed `Yoke::attach_to_borrowed_cart`, etc., to a more general `Yoke::attach_to_zero_copy_cart`

I was also thinking of renaming `Yoke::attach_to_cart` to be `Yoke::attach_to_cart_with` but I stopped here.

### CODE REVIEW

File: utils/yoke/src/either.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+//! Types to enable polymorphic carts.
+
+use crate::CloneableCart;
+use crate::Yoke;
+use crate::Yokeable;
+use core::ops::Deref;
+use stable_deref_trait::StableDeref;
+
+/// A cart that can be one type or the other. Enables ergonomic polymorphic carts.
+///
```
[the remark this answers is not in the document]
Comment (sffc) [reply utils/yoke/src/either.rs:14]: I added a note in the Yoke docs.  I don't think EitherCart is necessarily the most appropriate place to put that note.  We also have a paragraph talking about the destructor in the erase function docs.

File: utils/yoke/src/yoke.rs [hunk 2]
```diff
@@ -28,7 +28,9 @@ use alloc::sync::Arc;
 /// not the actual lifetime of the data, rather it is a convenient way to erase
 /// the lifetime and make it dynamic.
 ///
-/// `C` is the "cart", which `Y` may contain references to.
+/// `C` is the "cart", which `Y` may contain references to. After the yoke is constructed,
+/// the cart serves little purpose except to guarantee that `Y`'s references remain valid
+/// for as long as the yoke remains in memory.
```
Comment (Manishearth) [inline utils/yoke/src/yoke.rs:33]:
```suggestion
/// for as long as the yoke remains in memory (by calling the destructor at the appropriate moment).
```

**Review by Manishearth — APPROVED** [review]

File: utils/yoke/src/either.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply utils/yoke/src/either.rs:14]: In that case can we at least mention that the primary purpose is to be able to merge yokes of different carts?

[the remark this answers is not in the document]
Comment (sffc) [reply utils/yoke/src/either.rs:14]: done.

**Review by Manishearth — APPROVED** [review]

**Pull request merged**